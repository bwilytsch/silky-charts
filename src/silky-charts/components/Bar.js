import React, { useState, useRef, useEffect, useMemo } from 'react';
import identity from 'ramda/src/identity';
import { select as d3Select } from 'd3-selection';
import { max as d3Max } from 'd3-array';
import { axisBottom as d3AxisBottom, axisLeft as d3AxisLeft } from 'd3-axis';
import { Axis, BarDatum, Grid, SVG } from './styled';
import {
  allDate,
  drawGrid,
  extendXPath,
  getBaseColor,
  getDivergence,
  getId,
  getSize,
  getXScale,
  getYScale,
  rotateXLabels,
  setupData,
  valueFor,
} from '../utils';
import {
  SCALE_TIME,
  SCALE_BAND,
  SCALE_LINEAR,
  THEME,
  TICKS,
  MARGIN,
  ASPECT_RATIO,
  SIZE,
} from '../utils/constants';

const Bar = ({
  aspectRatio = ASPECT_RATIO,
  data: chartData,
  showValue,
  showDivergence,
  grid,
  horizontal,
  margin = MARGIN,
  onClick = identity,
  onMouseEnter = identity,
  onMouseLeave = identity,
  width: svgWidth = undefined,
  height: svgHeight = undefined,
  theme = THEME,
  ticks = TICKS,
  xAxisLabelRotation,
  xAxisLabelRotationValue = -50,
}) => {
  const svgRef = useRef();
  const [{ width, height, isSizeSet }, setSize] = useState(SIZE);
  const [currentValue, setCurrentValue] = useState(null);
  const [isDates, data] = useMemo(() => setupData(chartData), chartData);
  const [id] = useState(getId('bar'));

  const xScale = getXScale(
    isDates ? SCALE_TIME : SCALE_BAND,
    data,
    width,
    true
  );
  const yScale = getYScale(
    SCALE_LINEAR,
    d3Max(data, ({ value }) => value),
    height
  );

  const handleOnMouseEnter = event => {
    setCurrentValue(event.target.getAttribute('value'));
    onMouseEnter(event);
  };

  const handleOnMouseLeave = event => {
    setCurrentValue(null);
    onMouseLeave(event);
  };

  const handleResize = () => {
    const offsetWidth = svgRef.current.parentElement.offsetWidth;
    if ((svgWidth || svgHeight) && !isSizeSet) {
      setSize({
        ...getSize(svgWidth, svgHeight, margin, aspectRatio),
        isSizeSet: true,
      });
    } else if (offsetWidth !== svgWidth - (margin.left + margin.right)) {
      setSize({
        ...getSize(offsetWidth, undefined, margin, aspectRatio),
        isSizeSet: true,
      });
    }
  };

  useEffect(() => {
    handleResize();
    window.addEventListener('resize', handleResize);

    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, []);

  return (
    <SVG
      ref={svgRef}
      identifier={id}
      size={{
        width: svgWidth || width + margin.left + margin.right,
        height: svgHeight || height + margin.top + margin.bottom,
      }}
    >
      <g
        className="container"
        transform={`translate(${margin.left}, ${margin.top})`}
      >
        {grid && (
          <Grid
            ref={node =>
              d3Select(node).call(
                drawGrid(horizontal, xScale, height, yScale, width, ticks)
              )
            }
          />
        )}

        <Axis
          axis="x"
          position={{ x: 0, y: height }}
          ref={node => {
            d3Select(node).call(d3AxisBottom(xScale));
            isDates && extendXPath(id, width);
            xAxisLabelRotation && rotateXLabels(id, xAxisLabelRotationValue);
          }}
        />
        <Axis
          axis="y"
          ref={node => d3Select(node).call(d3AxisLeft(yScale).ticks(ticks))}
        />

        <g className="data">
          {data.map(({ name, value }, idx) => (
            <BarDatum
              key={idx}
              showValue={showValue}
              datum={{
                name,
                value:
                  showValue && showDivergence && currentValue
                    ? getDivergence(value, currentValue)
                    : value,
              }}
              x={
                isDates
                  ? xScale(name) - valueFor('x', width, data.length)
                  : xScale(name)
              }
              y={yScale(value)}
              width={
                isDates
                  ? valueFor('width', width, data.length)
                  : xScale.bandwidth()
              }
              height={height - yScale(value)}
              color={getBaseColor(theme)}
              onClick={onClick}
              onMouseEnter={handleOnMouseEnter}
              onMouseLeave={handleOnMouseLeave}
            />
          ))}
        </g>
      </g>
    </SVG>
  );
};

export default Bar;
