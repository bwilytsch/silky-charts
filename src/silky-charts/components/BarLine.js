import React, { useState } from 'react';
import head from 'ramda/src/head';
import identity from 'ramda/src/identity';
import uuidv4 from 'uuid/v4';
import { select as d3Select } from 'd3-selection';
import { max as d3Max } from 'd3-array';
import { axisBottom as d3AxisBottom, axisLeft as d3AxisLeft } from 'd3-axis';
import { Axis, StackedBarDatum, Grid, LineDatum, SVG } from './styled';
import {
  allDate,
  buildLine,
  buildStack,
  drawGrid,
  extendXPath,
  getLineDataForKeys,
  getMaximumValues,
  getSize,
  getXScale,
  getYScale,
  palette,
  rotateXLabels,
} from '../utils';
import {
  SCALE_TIME,
  SCALE_BAND,
  SCALE_LINEAR,
  SECONDARY_THEME,
  THEME,
  TICKS,
  LINE_TYPE,
} from '../utils/constants';

const BarLine = ({
  data,
  grid,
  horizontal,
  margin = { top: 20, right: 20, bottom: 30, left: 40 },
  onClick = identity,
  onMouseEnter = identity,
  onMouseLeave = identity,
  stackedKeys = [],
  lineKeys = [],
  lineType = LINE_TYPE,
  lineTypeOption = null,
  width: svgWidth = 960,
  height: svgHeight = 540,
  theme = THEME,
  ticks = TICKS,
  secondaryTheme = SECONDARY_THEME,
  xAxisLabelRotation,
  xAxisLabelRotationValue = -50,
}) => {
  const [width, height] = getSize(svgWidth, svgHeight, margin);
  const isNamesDate = allDate(data.map(({ name }) => name));

  const [id] = useState(`bar-line-${uuidv4()}`);

  const x = getXScale(isNamesDate ? SCALE_TIME : SCALE_BAND, data, width);
  const y = getYScale(
    SCALE_LINEAR,
    d3Max(getMaximumValues(stackedKeys, data)),
    height
  );

  const stack = buildStack(stackedKeys);
  const line = buildLine(x, y, lineType, lineTypeOption);

  const lineData = getLineDataForKeys(lineKeys, data);

  return (
    <SVG identifier={id} size={{ width: svgWidth, height: svgHeight }}>
      <g
        className="container"
        transform={`translate(${margin.left}, ${margin.top})`}
      >
        {grid && (
          <Grid
            ref={node =>
              d3Select(node).call(
                drawGrid(horizontal, x, height, y, width, ticks)
              )
            }
          />
        )}

        <Axis
          axis="x"
          position={{ x: 0, y: height }}
          ref={node => {
            d3Select(node).call(d3AxisBottom(x));
            isNamesDate && extendXPath(id, width);
            xAxisLabelRotation && rotateXLabels(xAxisLabelRotationValue);
          }}
        />
        <Axis
          axis="y"
          ref={node => d3Select(node).call(d3AxisLeft(y).ticks(ticks))}
        />

        <StackedBarDatum
          data={data}
          series={stack(data)}
          onClick={onClick}
          theme={theme}
          x={x}
          y={y}
          width={width}
          height={height}
        />

        {lineData.map((datum, idx) => {
          console.log(datum);

          return (
            <g className={`${head(datum)['key']}-layer`} key={idx}>
              <LineDatum
                data={datum}
                d={line(datum)}
                xScale={x}
                yScale={y}
                color={palette.themes[secondaryTheme].base[idx]}
              />
            </g>
          );
        })}
      </g>
    </SVG>
  );
};

export default BarLine;
