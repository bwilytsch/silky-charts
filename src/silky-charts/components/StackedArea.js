import React, { useEffect, useMemo, useRef, useState } from 'react';
import identity from 'ramda/src/identity';
import { axisBottom as d3AxisBottom, axisLeft as d3AxisLeft } from 'd3-axis';
import { select as d3Select } from 'd3-selection';
import { area as d3Area } from 'd3-shape';
import { Axis, Grid, Label, Path, SVG } from './styled';
import {
  appendStackedValues,
  buildStack,
  bySeries,
  classify,
  debounce,
  drawGrid,
  getId,
  getMax,
  getStackedMax,
  getSeries,
  getSize,
  getXScale,
  getYScale,
  palette,
  rotateXLabels,
  setLineType,
  setupData,
  toStackedForm,
} from '../utils';
import {
  ASPECT_RATIO,
  LINE_TYPE,
  MARGIN,
  ROTATION,
  SCALE_BAND,
  SCALE_LINEAR,
  SCALE_TIME,
  SIZE,
  THEME,
  TICKS,
} from '../utils/constants';

const StackedArea = ({
  aspectRatio = ASPECT_RATIO,
  data: chartData,
  grid,
  height: svgHeight = undefined,
  isHorizontal,
  lineSeries = [],
  lineType = LINE_TYPE,
  lineTypeOption = null,
  margin = MARGIN,
  onClick = identity,
  onMouseEnter = identity,
  onMouseLeave = identity,
  responsive = false,
  theme = THEME,
  ticks = TICKS,
  width: svgWidth = undefined,
  xAxisLabel,
  xAxisLabelRotation,
  xAxisLabelRotationValue = ROTATION,
  yAxisLabel,
}) => {
  const svgRef = useRef();
  const [id] = useState(getId('stacked-area'));
  const [{ width, height, isSizeSet }, setSize] = useState(SIZE);
  let [isDates, data] = useMemo(() => setupData(chartData), chartData);
  data = useMemo(
    () =>
      appendStackedValues(
        buildStack(getSeries(data))(toStackedForm(data)),
        data
      ),
    data
  );

  const xScale = getXScale(isDates ? SCALE_TIME : SCALE_BAND, data, width);
  const yScale = getYScale(SCALE_LINEAR, getMax(getStackedMax(data)), height);

  const area = d3Area()
    .curve(setLineType(lineType, lineTypeOption))
    .x(({ name }) => xScale(name))
    .y0(({ stackedValues }) => yScale(stackedValues[0]))
    .y1(({ stackedValues }) => yScale(stackedValues[1]));

  const handleSize = () => {
    const offsetWidth = svgRef.current.parentElement.offsetWidth;
    if ((svgWidth || svgHeight) && !isSizeSet) {
      setSize({
        ...getSize(svgWidth, svgHeight, margin, aspectRatio),
        isSizeSet: true,
      });
    } else if (offsetWidth !== svgWidth - (margin.left + margin.right)) {
      setSize({
        ...getSize(offsetWidth, undefined, margin, aspectRatio),
        isSizeSet: true,
      });
    }
  };

  const handleResize = debounce(handleSize)();

  useEffect(() => {
    handleSize();
    responsive && window.addEventListener('resize', handleResize);

    return () => {
      responsive && window.removeEventListener('resize', handleResize);
    };
  }, []);

  return (
    <SVG
      identifier={id}
      size={{
        width: svgWidth || width + margin.left + margin.right,
        height: svgHeight || height + margin.top + margin.bottom,
      }}
      ref={svgRef}
    >
      <g
        className="silky-charts-container"
        transform={`translate(${margin.left}, ${margin.top})`}
      >
        {grid && (
          <Grid
            ref={node =>
              d3Select(node).call(
                drawGrid(isHorizontal, xScale, height, yScale, width, ticks)
              )
            }
          />
        )}

        {xAxisLabel && (
          <Label axis="x" margin={margin} width={width} height={height}>
            {xAxisLabel}
          </Label>
        )}

        {yAxisLabel && (
          <Label axis="y" margin={margin} width={width} height={height}>
            {yAxisLabel}
          </Label>
        )}

        {bySeries(data).map(([series, datum], idx) => (
          <g className={`${classify(series)}-layer`} key={idx}>
            <Path
              chart="stacked-area"
              fillColor={palette.themes[theme].base[idx]}
              d={area(datum)}
              strokeWidth={0}
              onClick={onClick}
              onMouseEnter={onMouseEnter}
              onMouseLeave={onMouseLeave}
            />
          </g>
        ))}

        <Axis
          axis="x"
          position={{ x: 0, y: height }}
          ref={node => {
            d3Select(node).call(d3AxisBottom(xScale));
            xAxisLabelRotation && rotateXLabels(id, xAxisLabelRotationValue);
          }}
        />
        <Axis
          axis="y"
          ref={node => d3Select(node).call(d3AxisLeft(yScale).ticks(ticks))}
        />
      </g>
    </SVG>
  );
};

export default StackedArea;
