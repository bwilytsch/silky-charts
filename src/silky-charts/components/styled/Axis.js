import styled from 'styled-components';

const Axis = styled.g.attrs(({ axis, position }) => ({
  className: `axis-${axis}`,
  transform: position && `translate(${position.x}, ${position.y})`,
}))``;

export default Axis;
