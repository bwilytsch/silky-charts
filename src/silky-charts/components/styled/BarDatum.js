import React, { Fragment } from 'react';
import { Rect, Text } from './';

const BarDatum = ({
  color,
  datum: { value },
  x,
  y,
  width,
  height,
  onClick,
  onMouseEnter,
  onMouseLeave,
  showValue,
}) => (
  <Fragment>
    <Rect
      chart="bar"
      onClick={onClick}
      onMouseEnter={onMouseEnter}
      onMouseLeave={onMouseLeave}
      position={{ x, y }}
      size={{ width, height }}
      color={color}
      value={value}
    />
    {showValue && (
      <Text position={{ x, y }} size={{ width, height }}>
        {value}
      </Text>
    )}
  </Fragment>
);

export default BarDatum;
