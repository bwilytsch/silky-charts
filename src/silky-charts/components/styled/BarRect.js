import React, { Fragment } from 'react';
import styled from 'styled-components';
import { bar, black, white } from '../../utils/themes';

const Rect = styled.rect.attrs(({ position: { x, y } }) => ({
  x,
  y,
  className: 'bar',
}))`
  fill: ${({ theme }) => bar[theme].bg};
  height: ${({ size }) => size.height}px;
  width: ${({ size }) => size.width}px;

  &:hover {
    fill: ${({ theme }) => bar[theme].fg};
  }
`;

const Text = styled.text.attrs(
  ({ position: { x, y }, size: { width, height } }) => ({
    className: 'value',
    textAnchor: 'middle',
    transform: `translate(${x + width / 2}, ${y + (height < 40 ? -18 : 30)})`,
  })
)`
  fill: ${({ size: { height } }) => (height < 40 ? black : white)};
  pointer-events: none;
`;

const BarRect = ({
  datum: { name, value },
  position,
  x,
  y,
  width,
  height,
  onClick,
  onMouseEnter,
  onMouseLeave,
  showValue,
  theme,
}) => (
  <Fragment>
    <Rect
      position={{ x, y }}
      size={{ width, height }}
      theme={theme}
      onClick={onClick}
      onMouseEnter={onMouseEnter}
      onMouseLeave={onMouseLeave}
    >
      {value}
    </Rect>
    {showValue && (
      <Text position={{ x, y }} size={{ width, height }}>
        {value}
      </Text>
    )}
  </Fragment>
);

export default BarRect;
