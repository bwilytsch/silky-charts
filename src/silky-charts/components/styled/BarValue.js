import styled from 'styled-components';
import { black, white } from '../../utils/palette';

const BarValue = styled.text.attrs(
  ({ position: { x, y }, size: { width, height } }) => ({
    className: 'value',
    textAnchor: 'middle',
    transform: `translate(${x + width / 2}, ${y + (height < 40 ? -18 : 30)})`,
  })
)`
  fill: ${({ size: { height } }) => (height < 40 ? black : white)};
  pointer-events: none;
`;

export default BarValue;
