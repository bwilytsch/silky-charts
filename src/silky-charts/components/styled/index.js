export { default as Axis } from './Axis';
export { default as BarDatum } from './BarDatum';
export { default as Circle } from './Circle';
export { default as Grid } from './Grid';
export { default as LineDatum } from './LineDatum';
export { default as Path } from './Path';
export { default as Rect } from './Rect';
export { default as StackedBarDatum } from './StackedBarDatum';
export { default as SVG } from './SVG';
export { default as Text } from './Text';
