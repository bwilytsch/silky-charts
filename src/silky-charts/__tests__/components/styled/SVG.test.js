import React from 'react';
import { SVG } from '../../../components/styled';
import { create } from 'react-test-renderer';
import 'jest-styled-components';

test('render correctly', () => {
  const ref = jest.fn();
  const tree = create(
    <SVG identifier="foo" size={{ width: 10, height: 10 }} ref={ref} />
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
