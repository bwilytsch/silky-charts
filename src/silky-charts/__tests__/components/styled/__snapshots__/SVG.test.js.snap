// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render correctly 1`] = `
.c0 {
  height: 10px;
  width: 10px;
}

<svg
  className="silky-charts c0"
  id="foo"
  size={
    Object {
      "height": 10,
      "width": 10,
    }
  }
/>
`;
